crates/wasi/src/preview2/stream.rs

```diff
@@ -151,60 +133,354 @@ pub trait OutputStream: Send + Sync {
+            // Make a synchronous, non-blocking call attempt to read. We are not
+            // going to poll this more than once, so the noop waker is appropriate.
+            match Pin::new(&mut self.reader).poll_read(&mut cx, &mut readbuf) {
+                Poll::Pending => {}             // Nothing was read
+                Poll::Ready(result) => result?, // Maybe an error occured
+            };
+            let bytes_read = readbuf.filled().len();
+
+            if bytes_read == 0 {
+                self.state = StreamState::Closed;
+            }
+            bytes_read
+        } else {
+            0
+        };
+
+        Ok(((l + rest) as u64, self.state))
+    }
+
+    async fn ready(&mut self) -> Result<(), Error> {
+        if self.state.is_closed() {
+            return Ok(());
+        }
+
+        let mut bytes = core::mem::take(&mut self.buffer);
+        let start = bytes.len();
+        bytes.resize(start + 1024, 0);
+        let l =
+            tokio::io::AsyncReadExt::read_buf(&mut self.reader, &mut &mut bytes[start..]).await?;
+
+        // Reading 0 bytes means either there wasn't enough space in the buffer (which we
+        // know there is because we just resized) or that the stream has closed. Thus, we
+        // know the stream has closed here.
+        if l == 0 {
+            self.state = StreamState::Closed;
+        }
+
+        bytes.drain(start + l..);
+        self.buffer = bytes;
+
+        Ok(())
+    }
+}
+
+/// Provides a [`HostOutputStream`] impl from a [`tokio::io::AsyncWrite`] impl
+pub struct AsyncWriteStream<T> {
+    buffer: Vec<u8>,
+    writer: T,
+}
+
+impl<T> AsyncWriteStream<T> {
+    /// Create a [`AsyncWriteStream`]. In order to use the [`HostOutputStream`] impl
+    /// provided by this struct, the argument must impl [`tokio::io::AsyncWrite`].
+    pub fn new(writer: T) -> Self {
+        AsyncWriteStream {
+            buffer: Vec::new(),
+            writer,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<T: tokio::io::AsyncWrite + Send + Sync + Unpin + 'static> HostOutputStream
+    for AsyncWriteStream<T>
+{
+    // I can get rid of the `async` here once the lock is no longer a tokio lock:
+    fn write(&mut self, buf: &[u8]) -> Result<u64, anyhow::Error> {
+        let mut bytes = core::mem::take(&mut self.buffer);
+        bytes.extend(buf);
+
+        let noop_waker = noop_waker();
+        let mut cx: Context<'_> = Context::from_waker(&noop_waker);
+        // Make a synchronous, non-blocking call attempt to write. We are not
+        // going to poll this more than once, so the noop waker is appropriate.
+        match Pin::new(&mut self.writer).poll_write(&mut cx, &mut bytes.as_slice()) {
+            Poll::Pending => {
+                // Nothing was written: buffer all of it below.
+            }
+            Poll::Ready(written) => {
+                // So much was written:
+                bytes.drain(..written?);
```

Comment:
The error here is tricky to handle because data has been buffered up from the beginning of the function but this is also reporting an error, so if the application re-wrote the data then it'd be written twice